### DISCUSSION TITLE
How to generate an extraction rules file (e.g. rcx_patterns.rules) for a modified tech lef?

### DISCUSSION THREAD

**Discussion by xiaosiCU** [The-OpenROAD-Project/OpenROAD#6218] [Q&A] [answered]
I am trying to add several routing metal layers for the Nangate45 technology LEF, but during RC extraction, I need to use a pre-generated `rcx_patterns.rules` file. I followed the steps in the tool's README, but I do not have a golden SPEF or a reference extractor. If I skip this step, i.e.:
```
read_lef modified_nangate45.tech.lef
bench_wires -len 100 -all
write_rules -file rcx_patterns.rules
```
The rules file I generate has the correct format, but all the values are 0, meaning no valid information is extracted. What should I do?

**Comment by xiaosiCU** [comment]
I have tried the repo https://github.com/efabless/openrcx-calibration, through which I succeeded to generate valuable rcx_patterns.rules for PDK sky130. 

However, the repo provided the necessary sky130A.magicrc to launch the magic tool. I cannot find such a configuration for Nangate45 PDK. 

Could you please introduce in detail that how the platform/nangate45/rcx_patterns.rules was generated?

**Comment by maliberty** [comment, marked as answer]
Without a golden spef there is no way to generate a valid rules file.  As ng45 is non-manufacturable you can invent whatever values serve your purpose.

**Reply by xiaosiCU** [reply]
I have used the magic.tech provided by the ORFS ng45 pdk, successfully extracted the ext with Magic, and converted it into a golden spef, obtaining valid rules. Could you provide some information to help me create a custom magic.tech file? How do I convert from tlef?

Thank you for your reply!

**Reply by maliberty** [reply]
This is for OR issues not magic.

**Reply by xiaosiCU** [reply]
Thanks again!
